distributed/tests/test_scheduler.py

```diff
@@ -2925,14 +2926,196 @@ async def test_rebalance_least_recently_inserted_sender_min(c, s, *_):
+    x = await c.scatter("x", workers=[a.address])
+    bad_addr = "tcp://127.0.0.1:12345"
+    # Order matters; test both
+    addrs = [bad_addr, a.address] if missing_first else [a.address, bad_addr]
+    out = await s._gather_on_worker(b.address, {x.key: addrs})
+    assert out == set()
+    assert a.data[x.key] == "x"
+    assert b.data[x.key] == "x"
+
+
+@gen_cluster(client=True)
+async def test_gather_on_worker_key_not_on_sender(c, s, a, b):
+    """The only sender for a key does not actually hold it"""
+    out = await s._gather_on_worker(a.address, {"x": [b.address]})
+    assert out == {"x"}
+
+
+@pytest.mark.parametrize("missing_first", [False, True])
+@gen_cluster(client=True, nthreads=[("127.0.0.1", 1)] * 3)
+async def test_gather_on_worker_key_not_on_sender_replicated(
+    client, s, a, b, c, missing_first
+):
+    """One of the senders for a key does not actually hold it, but the key is available
+    somewhere else
+    """
+    x = await client.scatter("x", workers=[a.address])
+    # Order matters; test both
+    addrs = [b.address, a.address] if missing_first else [a.address, b.address]
+    out = await s._gather_on_worker(c.address, {x.key: addrs})
+    assert out == set()
+    assert a.data[x.key] == "x"
+    assert c.data[x.key] == "x"
+
+
+@gen_cluster(client=True, nthreads=[("127.0.0.1", 1)] * 3)
+async def test_gather_on_worker_duplicate_task(client, s, a, b, c):
+    """Race condition where the recipient worker receives the same task twice.
+    Test that the task nbytes are not double-counted on the recipient.
+    """
+    x = await client.scatter("x", workers=[a.address, b.address], broadcast=True)
+    assert a.data[x.key] == "x"
+    assert b.data[x.key] == "x"
+    assert x.key not in c.data
+
+    out = await asyncio.gather(
+        s._gather_on_worker(c.address, {x.key: [a.address]}),
+        s._gather_on_worker(c.address, {x.key: [b.address]}),
+    )
+    assert out == [set(), set()]
+    assert c.data[x.key] == "x"
+
+    a_ws = s.workers[a.address]
+    b_ws = s.workers[b.address]
+    c_ws = s.workers[c.address]
+    assert a_ws.nbytes > 0
+    assert c_ws.nbytes == b_ws.nbytes == a_ws.nbytes
+
+
+@gen_cluster(
+    client=True, nthreads=[("127.0.0.1", 1)] * 3, scheduler_kwargs={"timeout": "100ms"}
+)
+async def test_rebalance_dead_recipient(client, s, a, b, c):
+    """A key fails to be rebalanced due to recipient failure.
+    The key is not deleted from the sender.
+    Unrelated, successful keys are deleted from the senders.
+    """
+    x, y = await client.scatter(["x", "y"], workers=[a.address])
+    a_ws = s.workers[a.address]
+    b_ws = s.workers[b.address]
+    c_ws = s.workers[c.address]
+    x_ts = s.tasks[x.key]
+    y_ts = s.tasks[y.key]
+    await c.close()
+    assert s.workers.keys() == {a.address, b.address}
+
+    out = await s._rebalance_move_data([(a_ws, b_ws, x_ts), (a_ws, c_ws, y_ts)])
+    # FIXME status code is misleading
```

Comment:
You intend to keep the status code for compat reasons or is this still a TODO?

Reply:
Renamed missing-data to partial-fail